sonic_platform_base/sonic_xcvr/xcvr_api_factory.py

```diff
@@ -0,0 +1,28 @@
+"""
+    xcvr_api_factory.py
+
+    Factory class responsible for instantiating the appropriate XcvrApi
+    implementation for a xcvr module in SONiC
+"""
+
+from .xcvr_eeprom import XcvrEeprom
+# TODO: remove the following imports
+from .codes.public.sff8024 import Sff8024
+from .api.public.cmis import CmisApi
+from .mem_maps.public.cmis import CmisMemMap
+
+class XcvrApiFactory(object):
+    def __init__(self, reader, writer):
+        self.reader = reader
+        self.writer = writer
+
+    def _get_id(self):
+        # TODO: read ID from eeprom
+        pass
+
+    def create_xcvr_api(self):
+        # TODO: load correct classes from id_mapping file
+        codes = Sff8024
```

Comment:
In addition to Sff8024 where we use XCVR_IDENTIFIERS, we will possibly need these dictionaries in the codes too.

1. MODULE_STATUS_DICT = {
        1: 'ModuleLowPwr',
        2: 'ModulePwrUp',
        3: 'ModuleReady',
        4: 'ModulePwrDn',
        5: 'Fault'
    }

2. DATA_PATH_STATUS_DICT = {
        1: 'DataPathDeactivated',
        2: 'DataPathInit',
        3: 'DataPathDeinit',
        4: 'DataPathActivated',
        5: 'DataPathTxTurnOn',
        6: 'DataPathTxTurnOff',
        7: 'DataPathInitialized'
    }

3. LASER_TUNING_SETTING_DICT = {
        0: '3.125 GHz',
        1: '6.25 GHz',
        2: '12.5 GHz',
        3: '25 GHz',
        4: '50 GHz',
        5: '100 GHz',
        6: '33 GHz',
        7: '75 GHz'
    }

4. LASER_TUNING_STATUS_DICT = {
        0: 'Tuning completes; wavelength locked',
        1: 'Tuning not in progress; wavelength unlocked',
        3: 'Tuning in progress'
    }

5.  VDM_TYPE = {
        # VDM_ID: [VDM_NAME, DATA_TYPE, SCALE]
        1: ['Laser Age [%]', 'U16', 1],
        2: ['TEC Current [%]', 'S16', 100.0/32767],
        3: ['Laser Frequency Error [MHz]', 'S16', 10],
        4: ['Laser Temperature [C]', 'S16', 1.0/256],
        5: ['eSNR Media Input [dB]', 'U16', 1.0/256],
        6: ['eSNR Host Input [dB]', 'U16', 1.0/256],
        7: ['PAM4 Level Transition Parameter Media Input [dB]', 'U16', 1.0/256],
        8: ['PAM4 Level Transition Parameter Host Input [dB]', 'U16', 1.0/256],
        9: ['Pre-FEC BER Minimum Media Input', 'F16', 1], 
        10: ['Pre-FEC BER Minimum Host Input', 'F16', 1], 
        11: ['Pre-FEC BER Maximum Media Input', 'F16', 1], 
        12: ['Pre-FEC BER Maximum Host Input', 'F16', 1], 
        13: ['Pre-FEC BER Average Media Input', 'F16', 1], 
        14: ['Pre-FEC BER Average Host Input', 'F16', 1], 
        15: ['Pre-FEC BER Current Value Media Input', 'F16', 1],
        16: ['Pre-FEC BER Current Value Host Input', 'F16', 1],
        17: ['Errored Frames Minimum Media Input', 'F16', 1], 
        18: ['Errored Frames Minimum Host Input', 'F16', 1], 
        19: ['Errored Frames Maximum Media Input', 'F16', 1], 
        20: ['Errored Frames Minimum Host Input', 'F16', 1], 
        21: ['Errored Frames Average Media Input', 'F16', 1], 
        22: ['Errored Frames Average Host Input', 'F16', 1], 
        23: ['Errored Frames Current Value Media Input', 'F16', 1], 
        24: ['Errored Frames Current Value Host Input', 'F16', 1],
        128: ['Modulator Bias X/I [%]', 'U16', 100.0/65535],
        129: ['Modulator Bias X/Q [%]', 'U16', 100.0/65535],
        130: ['Modulator Bias Y/I [%]', 'U16', 100.0/65535],
        131: ['Modulator Bias Y/Q [%]', 'U16', 100.0/65535],
        132: ['Modulator Bias X_Phase [%]', 'U16', 100.0/65535],
        133: ['Modulator Bias Y_Phase [%]', 'U16', 100.0/65535],
        134: ['CD high granularity, short link [ps/nm]', 'S16', 1], 
        135: ['CD low granularity, long link [ps/nm]', 'S16', 20],
        136: ['DGD [ps]', 'U16', 0.01],
        137: ['SOPMD [ps^2]', 'U16', 0.01],
        138: ['PDL [dB]', 'U16', 0.1],
        139: ['OSNR [dB]', 'U16', 0.1],
        140: ['eSNR [dB]', 'U16', 0.1],
        141: ['CFO [MHz]', 'S16', 1],
        142: ['EVM_modem [%]', 'U16', 100.0/65535],
        143: ['Tx Power [dBm]', 'S16', 0.01],
        144: ['Rx Total Power [dBm]', 'S16', 0.01],
        145: ['Rx Signal Power [dBm]', 'S16', 0.01],
        146: ['SOP ROC [krad/s]', 'U16', 1],
        147: ['MER [dB]', 'U16', 0.1]
    }

Reply:
As we discussed, the Sff8024 class will correspond to common codes listed in the SFF-8024 document. Dictionaries like these are CMIS-specific and would need to go into CMIS-specific XcvrCodes class that inherits from Sff8024. This CMIS-specific class would then need to be specified under the "codes" entry in the id_mapping file. I'll update the HLD regarding this.